### PR TITLE
Feature/revocation

### PR DESCRIPTION
**Pull request by gunnarif**
Revocation implementation. Also the API for the value sets

### CODE REVIEW

**Comment by gunnarif** [comment]
Hi @ryanbnl 
Can you review this?

**Comment by ryanbnl** [comment]
Nice! Sure!

**Comment by ryanbnl** [comment]
I'm going to dig in, so will hopefully be finished tomorrow :)

File: README.md [hunk 1]
```diff
@@ -26,8 +26,19 @@ var cwt = decoder.Decode(barcodeContent);
...
+// To check if certificate has been revoked
+RevocationVerifier revocationVerifier = new RevocationVerifier(revokationRepository);
+var isRevoked = revocationVerifier.IsRevoked(cwt);
 ```
 
+This library doesn't implement the secratarieatService which is a repository to host all of the issuers public keys.
+Also, the library doesn't implement a procedure to fetch certificates and verify that they are verified against the CSCA.
+The GatewayService does provide an interface to fetch the certificates
+
+The library dons not implement the revokationRepostiory that host the list of all revocation or a way to check if the hash of the certificate is revoked.
```
Comment (ryanbnl) [inline README.md]: revocationRepository

File: README.md [hunk 2]
```diff
@@ -80,6 +91,54 @@ using var filestream = File.Create("qrCode.png");
...
+// Get the upload certificate with a private key
+// From a certificate store (HSM) or this could also be a file or some other methods
+X509Certificate2 uploadCert = new X509Certificate2([ByteArrayOfTheCert])
+
+await gateway.UploadNewRevokationBatch(new GatewayRevocationBatch
+{
+    country = "IS",
+    expires = DateTime.Now.AddDays(10),
+    hashType = "UCI",
+    kid = "bIwe3F4lAk4=",
```
Comment (ryanbnl) [inline README.md]: This can be derived from the certificates btw :)

File: dgc/Gateway/GatewayDSCResponse.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace DCC.Gateway
+{
+    public class GatewayDSCResponse
+    {
+        public string kid { get; set; }
+        public DateTime timestamp { get; set; }
```
Comment (ryanbnl) [inline dgc/Gateway/GatewayDSCResponse.cs]: I would personally use the C# standard PascalCase for the property names and either configure the JSON serializer to use camalCase or add the attributes.

File: dgc/Gateway/GatewayRevocationBatch.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace DCC.Gateway
+{
+    public class GatewayRevocationBatch
+    {
+        public string country { get; set; }
+        public DateTime expires { get; set; }
+        public string kid { get; set; }
+        public string hashType { get; set; }
+        public List<GatewayRevocationBatchEntry> entries { get; set; }
```
Comment (ryanbnl) [inline dgc/Gateway/GatewayRevocationBatch.cs]: Generally better to return interfaces over implementations (IList<T> > List<T>).

File: dgc/Gateway/GatewayService.cs [hunk 5]
```diff
@@ -62,20 +69,154 @@ public async Task<GatewayDSCResponse[]> GetAllDscFromGateway()
...
+        {
+            string valuesseturl = "/valuesets/" + valuesetId;
+            var response = await HttpClient.GetAsync(valuesseturl);
+
+            response.EnsureSuccessStatusCode();
+
+            var content = await response.Content.ReadAsStringAsync();
+
+            var valueset = new Dictionary<string, GatewayValueset>();
+            using (var reader = new JsonTextReader(new System.IO.StringReader(content)))
```
Comment (ryanbnl) [inline dgc/Gateway/GatewayService.cs:107]: You could also deserialize the json using a deserializer (newtonsoft / inbuilt), leads to less code here I believe?

File: dgc/RevocationUtils.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Security.Cryptography;
+using System.Text;
+
+namespace DCC
+{
+    public static class RevocationUtils
```
Comment (ryanbnl) [inline dgc/RevocationUtils.cs:6]: The hash is computed over entire signature / uci / uci+country. Then only the first 16 bytes (128 bits) of the signature are used.

File: dgc/RevocationUtils.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        public static byte[] ComputeUCIHash(CWT cwt)
+        {
+            string uci = "";
+
+            if (cwt.DGCv1.Vaccination.Length > 0) uci = cwt.DGCv1.Vaccination[0].CertificateIdentifier;
+            else if (cwt.DGCv1.Recovery.Length > 0) uci = cwt.DGCv1.Recovery[0].CertificateIdentifier;
+            else if (cwt.DGCv1.Test.Length > 0) uci = cwt.DGCv1.Test[0].CertificateIdentifier;
+
+            var sha256 = SHA256.Create();
+            return sha256.ComputeHash(Encoding.UTF8.GetBytes(uci)[0..16]);
```
Comment (ryanbnl) [inline dgc/RevocationUtils.cs]: sha256.ComputeHash(Encoding.UTF8.GetBytes(uci))[0..16];

File: dgc/RevocationUtils.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        public static byte[] ComputeContryCodeUCIHash(CWT cwt)
+        {
+            string uci = "";
+
+            if (cwt.DGCv1.Vaccination.Length > 0) uci = cwt.DGCv1.Vaccination[0].CountryOfVaccination + cwt.DGCv1.Vaccination[0].CertificateIdentifier;
+            else if (cwt.DGCv1.Recovery.Length > 0) uci = cwt.DGCv1.Recovery[0].CountryOfTest + cwt.DGCv1.Recovery[0].CertificateIdentifier;
+            else if (cwt.DGCv1.Test.Length > 0) uci = cwt.DGCv1.Test[0].CountryOfTest + cwt.DGCv1.Test[0].CertificateIdentifier;
+
+            var sha256 = SHA256.Create();
+            return sha256.ComputeHash(Encoding.UTF8.GetBytes(uci)[0..16]);
```
Comment (ryanbnl) [inline dgc/RevocationUtils.cs]: sha256.ComputeHash(Encoding.UTF8.GetBytes(uci))[0..16];

File: dgc/RevocationUtils.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        public static byte[] ComputeSignatureHash(CWT cwt)
+        {
+            var sha256 = SHA256.Create();
+            var signlength = cwt.CoseMessage.Signature.Length;
+            return sha256.ComputeHash(cwt.CoseMessage.Signature[0..(signlength/2)])[0..16];
```
Comment (ryanbnl) [inline dgc/RevocationUtils.cs]: sha256.ComputeHash(cwt.CoseMessage.Signature)[0..16];

Comment (gunnarif) [reply dgc/RevocationUtils.cs]: When I look at this from TSystem https://github.com/eu-digital-green-certificates/dgca-verifier-app-android/blob/ad25372a743c1c50aea4f893ad0d9cfbeee001a8/revocation/src/main/java/dcc/app/revocation/domain/Util.kt#L44
And the only limited test data that I could use, this works on that.

Comment (gunnarif) [reply dgc/RevocationUtils.cs]: But i need to add the RSA case to this

Comment (ryanbnl) [reply dgc/RevocationUtils.cs]: Aahh so R is simply the first 16 bytes? Handy. I'd recommend documenting it :-)

Comment (ryanbnl) [reply dgc/RevocationUtils.cs]: Specs are here btw: https://docs.google.com/document/d/1ozkuF8awPQQxVKH8Vx_03yRgAA-MYzb-9b_6E7Xqhlc/edit

Comment (ryanbnl) [reply dgc/RevocationUtils.cs]: @gunnarif think that this should work: [feature/revocation-with-rsa-signature-support](https://github.com/ehn-dcc-development/hcert-dotnet/tree/feature/revocation-with-rsa-signature-support); could do with adding some unit tests to cover it (can use the certs from the quality-assurance repository for that).

Comment (ryanbnl) [reply dgc/RevocationUtils.cs]: (to be more robust it should also throw if/when a new algorithm is encountered, wanted to keep the code a short as possible though)

File: dgc/RevocationUtils.cs (hunk 7 above)
Comment (ryanbnl) [reply dgc/RevocationUtils.cs]: btw seeing as you only use the SHA256 instance to compute the hash, you could just make one instance at class level with a static constructor or something + then reuse that. save allocations.

**Comment by gunnarif** [comment]
thanks @ryanbnl  
Think I addressed almost all of your comments. Except I had issues with the json sterilizer so I'm gonna leave it like this.

Did not find any good test data, only from DE from ES

**Review by ryanbnl — APPROVED** [review]
Looks good!

**Comment by ryanbnl** [comment]
Yeah not many countries have actually implemented Revocation yet. I guess that the political will / pressure has dropped for now.

**Pull request merged**